app_test.go

```diff
@@ -258,6 +262,75 @@ func TestNewApp(t *testing.T) {
 	})
 }
 
+type eventLogger struct {
+	l *zap.Logger
+}
+
+func (l *eventLogger) LogEvent(event fxevent.Event) {
+	l.l.Info("emitted event", zap.String("event", reflect.TypeOf(event).String()))
+}
+
+func setupFxLogger(l *zap.Logger) fxevent.Logger {
+	return &eventLogger{l: l}
+}
+
+func TestSetupLogger(t *testing.T) {
+	t.Parallel()
+	t.Run("initializing custom logger", func(t *testing.T) {
+		t.Parallel()
+		app := fxtest.New(t,
+			Provide(zap.NewExample),
+			WithLogger(setupFxLogger),
+		)
+		defer app.RequireStart().RequireStop()
+		require.NoError(t, app.Err())
+		require.NoError(t, app.Start(context.Background()))
+	})
+
+	t.Run("initializing zap production logger", func(t *testing.T) {
+		t.Parallel()
+		app := fxtest.New(t,
+			Provide(zap.NewProduction),
+			WithLogger(setupFxLogger),
+		)
+		defer app.RequireStart().RequireStop()
+		require.NoError(t, app.Err())
+		require.NoError(t, app.Start(context.Background()))
+	})
+
+	t.Run("initializing zap development logger", func(t *testing.T) {
+		t.Parallel()
+		app := fxtest.New(t,
+			Provide(zap.NewDevelopment),
+			WithLogger(setupFxLogger),
+		)
+		defer app.RequireStart().RequireStop()
+		require.NoError(t, app.Err())
+		require.NoError(t, app.Start(context.Background()))
+	})
+
+	t.Run("no custom logger", func(t *testing.T) {
+		t.Parallel()
+		app := fxtest.New(t)
+		defer app.RequireStart().RequireStop()
+		require.NoError(t, app.Err())
+		require.NoError(t, app.Start(context.Background()))
+	})
+
+	t.Run("error in WithLogger provider", func(t *testing.T) {
+		t.Parallel()
+		app := New(
+			WithLogger(&bytes.Buffer{}),
+		)
+		err := app.Err()
+		require.Error(t, err)
+		assert.Contains(t,
+			err.Error(),
+			"could not construct custom logger via fx.WithLogger: must provide constructor function, got  (type *bytes.Buffer)",
+		)
+	})
```

Comment:
Can we add a couple more test cases? In particular, we should test a logger with a dependency on other provided objects. This matches what we'll be using these APIs for. We should have these cases:

- Success
- Dependency failed to build because of an error
- Logger failed to build because of an error

We can test the output of the success case pretty easily: have the logger write to a buffer.
However, to test the output of the failure cases that write to stderr, we'll have to temporarily hijack os.Stderr by replacing it temporarily with an `os.File` for a temporary file (`ioutil.TempFile(t.TempDir(), "stderr")`?) and asserting its output.